### PR TITLE
Invalid cache before listing drive when collecting component names

### PR DESCRIPTION
**Pull request by ethanwharris**
## What does this PR do?

Can encounter a bug in the following setting (work a, work b, and a file server):
- work a - puts a file (creating a dir. under that works path)
- file server - gets the file (lists dirs and checks if any has the file, causing s3fs to cache the list of dirs, see: https://github.com/fsspec/s3fs/issues/181#issuecomment-488772002)
- work b - puts a file
- file server - tries to get the file but can't see it because of outdated cache

## Before submitting

- [ ] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

cc @borda

### CODE REVIEW

**Review by tchaton — APPROVED** [review]
Super neat !  Is it the only place ?

**Review by awaelchli — APPROVED** [review]

File: src/lightning_app/storage/drive.py [hunk 1]
```diff
@@ -231,6 +231,8 @@ def __deepcopy__(self, memo):
     def _collect_component_names(self) -> List[str]:
         sep = "/"
         if self.fs.exists(self.drive_root):
+            # Invalidate cache before running ls in case new directories have been added
+            self.fs.invalidate_cache()
```
Comment (hhsecond) [inline src/lightning_app/storage/drive.py:235]: do we know if this has any performance implications?

**Review by hhsecond — APPROVED** [review]

**Review by Borda — APPROVED** [review]

File: src/lightning_app/storage/drive.py (hunk 1 above)
Comment (lantiga) [reply src/lightning_app/storage/drive.py:235]: For sure this has implications with large nested fs layouts (which we don't have atm but they may come in the future, especially as we roll out training features). We don't have a workaround for now - the only one being monitoring changes incrementally (everytime we do an operation through the Drive API we update the cache accordingly).

@ethanwharris I would add a TODO here to re-evaluate.

**Review by lantiga — APPROVED** [review]

**Comment by Borda** [comment]
@ethanwharris mind check the comments so we can merge it :)

**Comment by Borda** [comment]
@ethanwharris, mind checking the remaining comments? :)

**Review by Borda — APPROVED** [review]

**Pull request merged**